### PR TITLE
Fix `compute_on_cpu` arg + forward method

### LINKED ISSUE #1323: `compute_on_cpu=True` not respected
## 🐛 Bug

I am using `torchmetrics.classification.roc.ROC` as a parent class in another project [1], and I had OOM issues because of the accumulation of preds and targes (when `threshold=None`). 

I tried using `compute_on_cpu=True` but the class doesn't seem to respect the option.

**Am I missing something?**
`preds.device` at training_step: cuda:0
self.auroc.preds[-1].device: cuda:0
self.auroc.compute().device: cuda:0
[1] The project is [`anomalib`](https://github.com/openvinotoolkit/anomalib), and the class inheriting from `ROC` is [`anomalib.utils.metrics.auroc.AUROC`](https://github.com/openvinotoolkit/anomalib/blob/74595c9a5a519cdc20fad1f3c3c68b5d0434cf44/anomalib/utils/metrics/auroc.py#L17) (yes, it shadows 'torchmetrics.AUROC').

---

I took screenshots of what's happenning inside the class `ROC` with vscode debugger. The training is done with pytorchlightning's trainer.

### `__init__()`

`self.compute_on_cpu` is correctly `True`.

debug init

### `update()`

Both `preds` (func arg) and `self.preds` (class attribute) are on the cpu, as expected.

Same for `target`/`self.target`.

debug update

### `compute()`

Both `preds` (func arg) and `self.preds` (class attribute) **were sent to the GPU** -- which I did **NOT** want.

Same for `target`/`self.target`.

debug compute

### To Reproduce

I got the [`mnist-hello-world.ipynb`](https://colab.research.google.com/github/PytorchLightning/lightning-tutorials/blob/publication/.notebooks/lightning_examples/mnist-hello-world.ipynb) from pytorchlightning's tutorials and modified it a bit to use that class in my project.

The notebook can be seen here: [github.com/jpcbertoldo/92eb5c91c4e8664bcc91ea34a4d9d37a](https://gist.github.com/jpcbertoldo/92eb5c91c4e8664bcc91ea34a4d9d37a).

The prints inside the `training_step()` show this (`FORCE_PREDS_TO_CPU = False`):

```text
`preds.device` at training_step: cuda:0
self.auroc.preds[-1].device: cuda:0
self.auroc.compute().device: cuda:0
```

I.e., the `compute_on_cpu` is not being respected -- right?

#### Code sample

Here is a simplified copy of `AUROC` from `anomalib`. I believe it's **not** influencing much because it's very simple and one doesn't see any `.to(device)`.

```py
class AUROC(ROC):
    """Area under the ROC curve."""

    def compute(self) -> Tensor:
        fpr, tpr = self._compute()
        # TODO: use stable sort after upgrading to pytorch 1.9.x (https://github.com/openvinotoolkit/anomalib/issues/92)
        if not (torch.all(fpr.diff() <= 0) or torch.all(fpr.diff() >= 0)):
            return auc(fpr, tpr, reorder=True)  # only reorder if fpr is not increasing or decreasing
        return auc(fpr, tpr)

    def update(self, preds: Tensor, target: Tensor) -> None:
        super().update(preds.flatten(), target.flatten())

    def _compute(self) -> Tuple[Tensor, Tensor]:
        fpr, tpr, _thresholds = super().compute()
        return (fpr, tpr)
```

### Expected behavior

I want the metric to be computed on the CPU and whatever internal tensor should stay in the CPU all times.

### Environment

Environment installed with miniconda, but the relevant dependencies are installed via pip (in the conda env file).

```
Python 3.8.13
pytorch-lightning             1.6.5
torch                         1.12.0
torchmetrics                  0.9.3
torchvision                   0.13.0
```

### PR DESCRIPTION
**Pull request by SkafteNicki**
## What does this PR do?

Fixes #1323

## Before submitting

- [x] Was this **discussed/approved** via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/metrics/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [ ] Did you make sure to **update the docs**?
- [x] Did you write any new **necessary tests**?

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Review by Borda — APPROVED** [review]

**Review by stancld — APPROVED** [review]

**Comment by jpcbertoldo** [comment]
Thanks @SkafteNicki!

**Comment by SkafteNicki** [comment]
@jpcbertoldo thanks for reporting such issues :]
